api/src/legacy_soap_api/legacy_soap_api_client.py

```diff
@@ -1,32 +1,95 @@
 
+logger = logging.getLogger(__name__)
 
-class LegacySOAPClient:
-    def __init__(self) -> None:
+
+class BaseSOAPClient:
+    def __init__(self, soap_request: SOAPRequest) -> None:
         self.config = LegacySoapAPIConfig()
+        self.soap_request = soap_request
+        self.soap_request_message = SoapPayload(self.soap_request.data.decode())
+        self.soap_request_operation_name = self.soap_request_message.operation_name
+        self.proxy_response = self._proxy_soap_request()
+        self.proxy_response_message = SoapPayload(self.proxy_response.data.decode())
 
-    def proxy_request(
-        self, method: str, full_path: str, headers: dict | None = None, body: bytes | None = None
-    ) -> LegacySOAPResponse:
+    def _proxy_soap_request(self) -> SOAPResponse:
         """Proxy incoming SOAP requests to grants.gov
         This method handles proxying requests to grants.gov SOAP API and retrieving
         and returning the xml data as is from the existing SOAP API.
         """
-        url = os.path.join(self.config.grants_gov_uri, full_path.lstrip("/"))
-        response = requests.request(method, url, data=body, headers=headers)
-        return LegacySOAPResponse(
+        response = requests.request(
+            method=self.soap_request.method,
+            url=os.path.join(self.config.grants_gov_uri, self.soap_request.full_path.lstrip("/")),
+            data=self.soap_request.data,
+            headers=self.soap_request.headers,
+        )
+        return SOAPResponse(
             data=self._process_response_response_content(response.content),
             status_code=response.status_code,
             headers=dict(response.headers),
         )
 
+    def get_request_soap_dict_body(self) -> dict:
+        return (
+            self.soap_request_message.to_dict()
+            .get("Envelope", {})
+            .get("Body", {})
+            .get(self.soap_request_operation_name, {})
+        )
+
     def _process_response_response_content(self, soap_content: bytes) -> bytes:
         if not self.config.inject_uuid_data:
             return soap_content
         return format_local_soap_response(soap_content)
+
+
+class SimplerApplicantsS2SClient(BaseSOAPClient):
+    def GetOpportunityListRequest(self) -> None:
+        get_opportunity_list_request = schemas.GetOpportunityListRequest(
+            **self.get_request_soap_dict_body()
+        )
+        # This will currently just log the validated schema until issue #4972 is implemented.
+        logger.info(
+            "soap get_opportunity_list_request validated",
+            extra={"get_opportunity_request": get_opportunity_list_request.model_dump()},
+        )
+        return None
+
+    def get_response(self) -> tuple:
+        # This method returns the raw response we get from the proxy request as well as
+        # the new simpler soap response data.
+        operation_method = self.__getattribute__(self.soap_request_operation_name)
```

Comment:
If you just want to see if the function is defined/fetch it, use `getattr(self, self.soap_request_operation_name, None)` instead. Same thing, but shorter / less internal function-y.